### PR TITLE
feat(scan): attribute app scans as device + app version instead of "Unknown device"

### LINKED ISSUE #2839: Scan records from the mobile app show "Unknown device"
## Current behavior

When a QR code is scanned with the companion app, the scan record's provenance panel on the asset page shows:

- Device: **Unknown device**
- Browser: **Unknown**
- OS: iOS (detected)

The app's API requests carry a generic user-agent, so the parser can't attribute the scan.

## Expected

Scans made with the app should read like: **Shelf app on iPhone (iOS 18)** — and ideally include the app version, which helps support tell whether a scan came from an outdated install.

## Approach

1. Companion: send an explicit identity on API requests from the shared client, e.g. `ShelfCompanion/1.3.0 (iPhone; iOS 18.6)`, built only from modules already in the shipped binary (react-native Platform + expo-constants) so the change stays OTA-eligible.
2. Webapp: recognize the `ShelfCompanion/` prefix wherever scan-record user agents are parsed for display, and render app name, device and app version instead of Unknown.

Older app versions keep showing Unknown until they update — acceptable.

## Verified on

Production scan record from app v1.3.0 (store build), 2026-08-12.

### PR DESCRIPTION
**Pull request by carlosvirreira**
Closes #2839.

## The problem

Scan a QR with the app and the asset page's scan panel says:

| | |
|---|---|
| Device | **Unknown device** |
| Browser | **Unknown** |

The app sets no `User-Agent`, so `ua-parser-js` has nothing to attribute. Verified on a production scan from the 1.3.0 store build.

## The change

**Companion** — `lib/api/user-agent.ts` builds one identity string, sent on every `apiFetch`/`apiUpload` request:

```
ShelfCompanion/1.3.0 (iPhone; iOS 18.6)
ShelfCompanion/1.3.0 (google Pixel 8; Android 14)
```

It uses **only** `react-native` `Platform` and `expo-constants` — both already compiled into the shipped binary (`EXConstants` is in `Podfile.lock`), so this needs **no new native build and is OTA-eligible**. Web builds send nothing, because browsers own that header.

**Webapp** — `parseCompanionUserAgent` recognizes that string in `parseScanData` and synthesizes the same shape `ua-parser-js` returns, so the panel renders **Apple - iPhone / Shelf app 1.3.0 / iOS** with no component changes. Any other UA falls through to `ua-parser-js` untouched.

The app version in the Browser row is deliberate: support can now tell at a glance whether a scan came from an outdated install.

## Notes

- Older app versions keep showing "Unknown device" until they update. Acceptable.
- `mobile-companion` (the server's null-header fallback in `qr.$qrId.ts`) still parses to Unknown by design — it means no UA arrived at all, and the test pins that.
- The feedback email's `formatUserAgent` needs no change: it already falls back to the raw string, which is now readable rather than absent.
- No new dependency, no schema change, no migration.

## Verification

- 9 new unit tests in `app/modules/scan/utils.server.test.ts` (iPhone, iPad, Android brand+model, bare-Android fallback, browser UAs, empty/null, and a look-alike that embeds the prefix mid-string).
- `tsc --noEmit` clean in both apps; ESLint clean on all four touched files.

## Summary by CodeRabbit

* **New Features**
  * Companion app requests now include app, device, and operating-system details when available.
  * Scan activity can identify and display companion app version and device information.
  * Standard browser and device detection remains supported as a fallback.

* **Bug Fixes**
  * Improved handling of missing, invalid, or unsupported client identification data.

* **Tests**
  * Added coverage for iPhone, iPad, Android, browser, and edge-case scan data.

### CODE REVIEW

File: apps/webapp/app/modules/scan/utils.server.test.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        "Mozilla/5.0 ShelfCompanion/1.3.0 (iPhone; iOS 18.6)"
+      )
+    ).toBeNull();
+  });
+});
+
+describe("parseScanData", () => {
+  // why: parseScanData only reads plain fields off the scan record; a
+  // literal fixture keeps the test honest without a database.
+  const baseScan = {
```
[the remark this answers is not in the document]
Comment (carlosvirreira) [reply apps/webapp/app/modules/scan/utils.server.test.ts]: Fixed in 108e7769b — fair catch, the double cast did defeat the point of typing the fixture.

- Exported `ScanWithRelations` from `modules/scan/utils.server` so the relation shape has a name, and `parseScanData` now binds to it.
- Added `test/factories/scan.ts` — `createScanWithRelations(overrides)` plus `createScanUser` / `createScanQr` for the relations — exported from `@factories`.
- The tests build fixtures from it with zero casts, so a `Scan` schema change now fails typecheck instead of silently exercising a stale shape. Added a third case pinning that the UA change leaves `coordinates` and `manuallyGenerated` untouched.

10 tests pass, typecheck and lint clean.

File: apps/webapp/test/factories/scan.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Factory for a `Scan` row carrying the relations `parseScanData` reads.
+ *
+ * Defaults to an anonymous companion-app scan with coordinates — the shape
+ * the mobile QR resolve route writes — so callers override only what their
+ * assertion is about (e.g. `userAgent`, or a `user` to test attribution).
+ *
+ * @param overrides - partial scan fields, including `user` and `qr` relations
+ * @returns a fully typed scan; a schema change breaks this at compile time
```
[the remark this answers is not in the document]
Comment (carlosvirreira) [reply apps/webapp/test/factories/scan.ts]: Fixed in f3c1ff0ca — right, and it was the same flaw the factory existed to remove: `as User` / `as Qr` hid the missing required columns (`User.tierId`, timestamps, `Qr.version`, `Qr.errorCorrection`).

Rather than build narrow projection types, I deleted both helpers: nothing used them. Every test builds its fixture from `createScanWithRelations`, whose `user` and `qr` default to `null`, and the remaining factory is fully typed with no assertions. A test that later needs those relations should pass fully-typed objects so a schema change fails the build instead of the assertion — noted in the JSDoc so the next person does not re-add a cast.

10 tests pass; typecheck and lint clean.

File: apps/webapp/app/modules/scan/utils.server.ts [hunk 3]
```diff
@@ -1,8 +1,74 @@
...
+  }
+
+  return {
+    ua: userAgent as string,
+    browser: {
+      name: `Shelf app ${appVersion}`,
+      version: appVersion,
+      major: undefined,
+    },
+    device: { vendor, model, type: "mobile" },
```
[the remark this answers is not in the document]
Comment (carlosvirreira) [reply apps/webapp/app/modules/scan/utils.server.ts]: Valid — fixed. `device.type` was hardcoded to `"mobile"`, so every iPad scan was attributed as a phone.

Now derived: `deviceToken === "iPad" ? "tablet" : "mobile"`, and the existing iPad test asserts the type rather than only the model.

I deliberately did not extend this to Android. Our user-agent carries a brand plus model token (`samsung SM-T870`) which does not state form factor, so classifying those would mean maintaining a device lookup table. Guessing there would produce confidently wrong attribution, which is worse than the honest default. Noted in a comment at the call site so the asymmetry does not read as an oversight.
